akka-typed/src/test/java/akka/typed/javadsl/Counter.java

```diff
@@ -0,0 +1,37 @@
+package akka.typed.javadsl;
+
+import akka.typed.ActorRef;
+import akka.typed.Behavior;
+
+public interface Counter {
+
+	public Counter incr();
+
+	public Counter get(ActorRef<Long> replyTo);
+
+	static class C implements Counter {
+		private final long c;
+
+		public C(long c) {
+			this.c = c;
+		}
+
+		@Override
+		public Counter incr() {
+			return new C(c + 1);
+		}
+
+		@Override
+		public Counter get(ActorRef<Long> replyTo) {
+			replyTo.tell(c);
+			return this;
+		}
+	}
+	
+	static final Behavior<State<Counter>> zero = new StateBehavior<>(new C(0));
+
+	public static Behavior<State<Counter>> initial() {
+		return zero;
+	}
+
+}
```

Comment:
There are four drawbacks:

* messages cannot reasonably be printed out since they are lambdas
* messages can be written such that they do complex compositions or other operations within the target Actor’s scope
* messages easily end up being non-serializable
* providing custom serializers or using them for “included” (i.e. captured) values will be tough

Given the amount of code needed to implement this (i.e. all of `State.scala`—four actual lines of code) it is trivial for anyone to put this into their projects, so I’m a little torn between hinting at this for custom implementation or documenting it carefully with the drawbacks. I am particularly worried about the lack of traceability that this will induce.

@pvlugter @bantonsson do you have opinions/insights? @ivantopo @dpsoft ?